File: Hard-problems/string-permutations.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace CodeEval
{
    class Program
    {
        static StringBuilder output = new StringBuilder();
        static void Main(string[] args)
        {
            using (StreamReader reader = File.OpenText(args[0]))
            {
                while (!reader.EndOfStream)
                {

                    string line = reader.ReadLine();
                    if (String.IsNullOrEmpty(line))
                        continue;
                    // do something with line

                    // custom character sort for problem: digits < upper case letters < lower case letters
                    char[] charsOfString = line.ToCharArray();
                    List<char> charList = charsOfString.Where(r => char.IsDigit(r)).OrderBy(r => r)
                        .Concat(charsOfString.Where(r => char.IsUpper(r)).OrderBy(r => r))
                        .Concat(charsOfString.Where(r => char.IsLower(r)).OrderBy(r => r)).ToList();

                    // input string now sorted according to problem's rules
                    string input = String.Join("", charList);

                    // clear the old output
                    output.Clear();

                    // perform the permutations
                    permutation(input);

                    output.Length--; // remove trailing comma

                    Console.WriteLine(output);
                }
                Console.ReadLine();
            }
        }

        public static void permutation(string input)
        {
            permutation("", input);
        }

        public static void permutation(string prefix, string suffix)
        {
            int n = suffix.Length;

            if (n == 0)
            {
                output.Append(prefix).Append(","); // ending condition
            }
            else 
            {
                for (int i = 0; i < n; i++)
                {
                    // prefix + the ith character of the suffix is the new prefix
                    // then permute the rest of the suffix without that ith character
                    permutation(prefix + suffix[i], suffix.Substring(0, i) + suffix.Substring(i + 1));
                }
            }
        }
    }
}

File: Hard-problems/string-permutations.java
import java.io.*;
import java.util.*;


public class Main {
	static StringBuilder output = new StringBuilder();
	
    public static void main (String[] args) throws IOException {
        File file = new File(args[0]);
        BufferedReader buffer = new BufferedReader(new FileReader(file));
        String line;
        while ((line = buffer.readLine()) != null) {
            line = line.trim();
            // Process line of input Here
            
            char[] charsOfString = line.toCharArray();
            // character sort for problem: digits < upper case letters < lower case letters
            // can use default .sort() for this
            Arrays.sort(charsOfString);
            
            // input string now sorted according to problem's rules
            String input = new String(charsOfString);

            // clear the old output
            output.setLength(0);
            
            // perform the permutations
            permutation(input);

            output.deleteCharAt(output.length() - 1); // remove trailing comma

            System.out.println(output);
            
        }
        
        buffer.close();
    }
    
    public static void permutation(String str) { 
        permutation("", str); 
    }

    private static void permutation(String prefix, String str) {
        int n = str.length();
        if (n == 0)
        {
        	output.append(prefix).append(",");
        }
        else {
            for (int i = 0; i < n; i++)
                permutation(prefix + str.charAt(i), str.substring(0, i) + str.substring(i+1, n));
        }
    }
    
}
